graph-subscriptions-api/src/subscriptions_subgraph.rs

```diff
@@ -0,0 +1,198 @@
+        let (subscriptions_tx, subscriptions_rx) = Eventual::new();
+        let client = Arc::new(Mutex::new(Client {
+            subgraph_client,
+            subscriptions: subscriptions_tx,
+        }));
+
+        eventuals::timer(Duration::from_secs(30))
+            .pipe_async(move |_| {
+                let client = client.clone();
+                async move {
+                    let mut client = client.lock().await;
+                    if let Err(poll_active_subscriptions_err) =
+                        client.poll_active_subscriptions().await
+                    {
+                        tracing::error!(%poll_active_subscriptions_err);
+                    }
+                }
+            })
+            .forever();
+
+        subscriptions_rx
+    }
+
+    async fn poll_active_subscriptions(&mut self) -> Result<(), String> {
+        // Serve queries for subscriptions that end 10 minutes ago and later.
+        let timestamp = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_millis() as u64;
+        let active_sub_end = (timestamp / 1000) - (60 * 10);
+
+        let query = format!(
+            r#"
+            activeSubscriptions(
+                block: $block
+                orderBy: id, orderDirection: asc
+                first: $first
+                where: {{
+                    id_gt: $last
+                    end_gte: {active_sub_end}
+                }}
+            ) {{
+                id
+                user {{
+                    id
+                    authorizedSigners {{
+                        signer
+                    }}
+                }}
+                start
+                end
+                rate
+            }}
+          "#,
+        );
+        let active_subscriptions_response = self
+            .subgraph_client
+            .paginated_query::<ActiveSubscription>(&query)
+            .await?;
+        if active_subscriptions_response.is_empty() {
+            return Err("Discarding empty update (active_subscriptions)".to_string());
+        }
+
+        let subscriptions_map = active_subscriptions_response
+            .into_iter()
+            .map(|active_sub| {
+                let user = active_sub.user;
+                let signers = user
+                    .authorized_signers
+                    .into_iter()
+                    .map(|signer| signer.signer)
+                    .chain([user.id]);
+                let sub = ActiveSubscriptionWithSigners {
+                    user: user.id,
+                    start: active_sub.start,
+                    end: active_sub.end,
+                    rate: active_sub.rate,
+                    signers: signers.collect(),
+                };
+                (user.id, sub)
+            })
+            .collect();
+        self.subscriptions.write(Ptr::new(subscriptions_map));
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Pretty rad. Is this the usual thing in Rust or do people usually create a new file?

Reply:
Same file for unit tests like this: https://doc.rust-lang.org/book/ch11-03-test-organization.html

Reply:
So it solves the question of the boundary between unit and integration tests. Nice.